pennylane/templates/subroutines/qdrift.py

```diff
@@ -0,0 +1,257 @@
+            raise ValueError(
+                "There should be atleast 2 terms in the Hamiltonian. Otherwise use `qml.exp`"
+            )
+
+        if any(requires_grad(coeff) for coeff in coeffs):
+            raise qml.QuantumFunctionError(
+                "The QDrift template currently doesn't support differentiation through the "
+                "coefficients of the input Hamiltonian."
+            )
+
+        if decomposition is None:  # need to do this to allow flatten and _unflatten
+            unwrapped_coeffs = unwrap(coeffs)
+            decomposition = _sample_decomposition(unwrapped_coeffs, ops, time, n=n, seed=seed)
+
+        self._hyperparameters = {
+            "n": n,
+            "seed": seed,
+            "base": hamiltonian,
+            "decomposition": decomposition,
+        }
+        super().__init__(time, wires=hamiltonian.wires, id=id)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        """Recreate an operation from its serialized format.
+
+        Args:
+            data: the trainable component of the operation
+            metadata: the non-trainable component of the operation
+
+        The output of ``Operator._flatten`` and the class type must be sufficient to reconstruct the original
+        operation with ``Operator._unflatten``.
+
+        **Example:**
+
+        >>> op = qml.Rot(1.2, 2.3, 3.4, wires=0)
+        >>> op._flatten()
+        ((1.2, 2.3, 3.4), (<Wires = [0]>, ()))
+        >>> qml.Rot._unflatten(*op._flatten())
+        >>> op = qml.PauliRot(1.2, "XY", wires=(0,1))
+        >>> op._flatten()
+        ((1.2,), (<Wires = [0, 1]>, (('pauli_word', 'XY'),)))
+        >>> op = qml.ctrl(qml.U2(3.4, 4.5, wires="a"), ("b", "c") )
+        >>> type(op)._unflatten(*op._flatten())
+        Controlled(U2(3.4, 4.5, wires=['a']), control_wires=['b', 'c'])
+
+        """
+        hyperparameters_dict = dict(metadata[1])
+        hamiltonian = hyperparameters_dict.pop("base")
+        return cls(hamiltonian, *data, **hyperparameters_dict)
+
+    @staticmethod
+    def compute_decomposition(*args, **kwargs):  # pylint: disable=unused-argument
+        r"""Representation of the operator as a product of other operators (static method).
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+        .. note::
+
+            Operations making up the decomposition should be queued within the
+            ``compute_decomposition`` method.
+
+        .. seealso:: :meth:`~.Operator.decomposition`.
+
+        Args:
+            *params (list): trainable parameters of the operator, as stored in the ``parameters`` attribute
+            wires (Iterable[Any], Wires): wires that the operator acts on
+            **hyperparams (dict): non-trainable hyperparameters of the operator, as stored in the ``hyperparameters`` attribute
+
+        Returns:
+            list[Operator]: decomposition of the operator
+        """
+        decomp = kwargs["decomposition"]
+
+        if qml.QueuingManager.recording():
```

Comment:
Why is this check done here? I don't see it done in other templates

Reply:
The `decomposition()` function is supposed to return and queue the decomposition operators. The `_sample_decomposition()` private method returns the correct decomposition but it makes a bunch of operators in the process so we have to put `stop_queuing()` around it. For this reason we need to manually queue each of the operators in the decomposition (if its called in a circuit)

Reply:
Got it! thanks 😊